compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -3661,64 +3683,69 @@ public void visit(BLangOnFailClause onFailClause) {
     }
 
     @Override
-    public void visit(BLangWhile whileNode) {
-        typeChecker.checkExpr(whileNode.expr, env, symTable.booleanType);
+    public void visit(BLangWhile whileNode, AnalyzerData data) {
+        SymbolEnv env = data.env;
+        typeChecker.checkExpr(whileNode.expr, env, symTable.booleanType, data.prevEnvs);
 
         if (whileNode.onFailClause != null) {
-            this.analyzeNode(whileNode.onFailClause, env);
+            this.analyzeNode(whileNode.onFailClause, data);
         }
 
         BType actualType = whileNode.expr.getBType();
         if (TypeTags.TUPLE == Types.getReferredType(actualType).tag) {
             dlog.error(whileNode.expr.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES, symTable.booleanType, actualType);
         }
 
-        boolean prevBreakFound = this.breakFound;
+        boolean prevBreakFound = data.breakFound;
         SymbolEnv whileEnv = typeNarrower.evaluateTruth(whileNode.expr, whileNode.body, env);
-        analyzeStmt(whileNode.body, whileEnv);
-        this.notCompletedNormally =
+        data.env = whileEnv;
+        analyzeStmt(whileNode.body, data);
+        data.notCompletedNormally =
                 ConditionResolver.checkConstCondition(types, symTable, whileNode.expr) == symTable.trueType
-                        && !this.breakFound;
-        this.breakFound = prevBreakFound;
+                        && !data.breakFound;
+        data.breakFound = prevBreakFound;
     }
 
     @Override
-    public void visit(BLangDo doNode) {
-        SymbolEnv narrowedEnv = SymbolEnv.createTypeNarrowedEnv(doNode, env);
+    public void visit(BLangDo doNode, AnalyzerData data) {
+        SymbolEnv narrowedEnv = SymbolEnv.createTypeNarrowedEnv(doNode, data.env);
         if (doNode.onFailClause != null) {
-            this.analyzeNode(doNode.onFailClause, narrowedEnv);
+            data.env = narrowedEnv;
+            this.analyzeNode(doNode.onFailClause, data);
         }
-        analyzeStmt(doNode.body, narrowedEnv);
+        data.env = narrowedEnv;
+        analyzeStmt(doNode.body, data);
```

Comment:
Can't we assign the new env to `data.env` at L:3711 and remove assignements at L:3713 and L:3716